keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+
+
+#### [Scoped for Beta] HugeTLB cgroup
+
+Note: This section includes only high level overview; Design details will be added in Beta stage.
+
+To support pod-level resource specifications for hugepages, Kubernetes will need to adjust how it handles hugetlb cgroups. Unlike memory, where an unset limit 
+means unlimited, an unset hugetlb limit is the same as setting it to 0.
+
+With the proposed changes, hugepages-2Mi and hugepages-1Gi will be added to the pod-level resources section, alongside CPU and memory. The hugetlb cgroup for the
+pod will then directly reflect the pod-level hugepage limits, rather than using an aggregated value from container limits. When scheduling, the scheduler will 
+consider hugepage requests at the pod level to find nodes with enough available resources.
+
+
+#### [Scoped for Beta] Topology Manager
+
+Note: This section includes only high level overview; Design details will be added in Beta stage.
+
+
+* (Tentative) Only pod level scope for topology alignment will be supported if pod level requests and limits are specified without container-level requests and limits.
+* The pod level scope for topology aligntment will consider pod level requests and limits instead of container level aggregates.
+* The hint providers will consider pod level requests and limits instead of container level aggregates.
+
+
+#### [Scoped for Beta] Memory Manager
+
+Note: This section includes only high level overview; Design details will be added in Beta stage.
+
+With the introduction of pod-level resource specifications, the Kubernetes Memory Manager will evolve to track and enforce resource limits at both the pod and container levels. It will need to aggregate memory usage across all containers within a pod to calculate the pod's total memory consumption. The Memory Manager will then enforce the pod-level limit as the hard cap for the entire pod's memory usage, preventing it from exceeding the allocated amount.  While still maintaining container-level limit enforcement, the Memory Manager will need to coordinate with the Kubelet and eviction manager to make decisions about pod eviction or individual container termination when the pod-level limit is breached.
+
+
+#### [Scoped for Beta] CPU Manager
+
+Note: This section includes only high level overview; Design details will be added in Beta stage.
+
+With the introduction of pod-level resource specifications, the CPU manager in Kubernetes will adapt to manage CPU requests and limits at the pod level rather than solely at the container level. This change means that the CPU manager will allocate and enforce CPU resources based on the total requirements of the entire pod, allowing for more flexible and efficient CPU utilization across all containers within a pod. The CPU manager will need to ensure that the aggregate CPU usage of all containers in a pod does not exceed the pod-level limits.
+
+#### [Scoped for Beta] In Place Pod Resize
+
+TBD. Do not review for the alpha stage.
```

Comment:
will we allow inplace to go into beta without pod level support while the pod level support is in alpha? Will we consider in place pod resize to require pod level support when both features are in alpha? 

If not - we will need to rename the kep to "in place containers resize" to highlight that pods are out of scope

Reply:
My thoughts: We shouldn't require a feature under development to support other alpha or even pre-alpha features. So in-place pod resizing going beta shouldn't be blocked on this feature. But better coordination/collaboration will help both.

I don't think we need to rename. We can just clearly call out the scope in the KEP. But @tallclair may have different opinions.

Reply:
getting both of those features to GA in a current scope will result in inconsistent user experience for sure.

Also small issues like inplace will need to account for cases when changing container's resources will put a Pod in inconsistent state (e.g. sum of requests > pod's request) needs to be accounted while in alpha/beta.

Since inplace is already in alpha, I think the validation I mentioned must be added by this KEP. Worth mentioning in this document

Reply:
> Also small issues like inplace will need to account for cases when changing container's resources will put a Pod in inconsistent state (e.g. sum of requests > pod's request) needs to be accounted while in alpha/beta.

In-place update simply provides a mechanism for users/components to update container resource specs. If the provided values are invalid, they'd be denied. We can further clarify the behavior in the KEP, but I think this is working as intended and is already in scope.